src/main/java/org/terracotta/offheapstore/paging/UpfrontAllocatingPageSource.java

```diff
@@ -440,67 +474,164 @@ public synchronized Runnable removeAllocationThreshold(ThresholdDirection direct
+        LOGGER.debug("Took {} ms to create off-heap storage of {}B.", TimeUnit.NANOSECONDS.toMillis(duration), DebuggingUtils.toBase2SuffixedString(toAllocate));
       }
-      LOGGER.debug("Allocated {}B in {}B chunks.", new Object[]{DebuggingUtils.toBase2SuffixedString(allocated - lastFailureAt), DebuggingUtils.toBase2SuffixedString(currentChunkSize)});
-      long duration = System.nanoTime() - start;
-      LOGGER.debug("Took {} ms to create off-heap storage of {}B.", TimeUnit.NANOSECONDS.toMillis(duration), DebuggingUtils.toBase2SuffixedString(max));
+
       return Collections.unmodifiableCollection(buffers);
     }
 
-    private static PrintStream createAllocatorLog(long max, int maxChunk, int minChunk) throws IOException {
+  private static Collection<ByteBuffer> bufferAllocation(BufferSource source, int toAllocate, int minChunk, boolean fixed, PrintStream allocatorLog, long start) {
+    long allocated = 0;
+    long currentChunkSize = toAllocate;
+
+    Collection<ByteBuffer> buffers = new ArrayList<ByteBuffer>();
+
+    while (allocated < toAllocate) {
+      long blockStart = System.nanoTime();
+      int currentAllocation = (int)Math.min(currentChunkSize, (toAllocate - allocated));
+      ByteBuffer b = source.allocateBuffer(currentAllocation);
+      long blockDuration = System.nanoTime() - blockStart;
+
+      if (b == null) {
+        if (fixed || (currentChunkSize >>> 1) < minChunk) {
+          throw new IllegalArgumentException("An attempt was made to allocate more off-heap memory than the JVM can allow." +
+                                             " The limit on off-heap memory size is given by the -XX:MaxDirectMemorySize command (or equivalent).");
+        }
+
+        // In case of failure, we try half the allocation size. It might pass if memory fragmentation caused the failure
+        currentChunkSize >>>= 1;
+
+        if(LOGGER.isDebugEnabled()) {
+          LOGGER.debug("Allocated failed at {}B, trying  {}B chunks.", DebuggingUtils.toBase2SuffixedString(currentAllocation), DebuggingUtils.toBase2SuffixedString(currentChunkSize));
+        }
+      } else {
+        buffers.add(b);
+        allocated += currentAllocation;
+
+        if (allocatorLog != null) {
+          allocatorLog.printf("%d,%d,%d,%d,%d,%d,%d,%d%n", System.nanoTime() - start,
+            Thread.currentThread().getId(),  blockDuration, currentAllocation, PhysicalMemory.freePhysicalMemory(), PhysicalMemory.totalSwapSpace(), PhysicalMemory.freeSwapSpace(), PhysicalMemory.ourCommittedVirtualMemory());
+        }
+
+        if(LOGGER.isDebugEnabled()) {
+          LOGGER.debug("{}B chunk allocated", DebuggingUtils.toBase2SuffixedString(currentAllocation));
+        }
+      }
+    }
+
+    return buffers;
+  }
+
+  private static <T> T uninterruptibleGet(Future<T> future, AtomicBoolean interrupted) {
+      while(true) {
+        try {
+          return future.get();
+        } catch (ExecutionException e) {
```

Comment:
I agree, the the `unterruptibleGet` is private and it makes the caller code much cleaner. So I prefer it here. I could wrap it in another method but that seems overkill